.github/workflows/build-test-ship.yaml

```diff
@@ -128,6 +146,143 @@ jobs:
+      # this step will be unnecessary.
+      - name: Use latest service images from ECR
+        run: sed -i 's/${USER:?USER}/${{ steps.login-ecr.outputs.registry }}/' docker-compose.yml
+
+      - name: Create docker-compose override and AWS credentials files
+        env:
+          ENCODED_CREDENTIALS: ${{ secrets.DEV_AWS_CREDENTIALS }}
+        shell: bash
+        run: |
+          echo -e "version: '3.7'\nservices:" >> docker-compose.ci.yaml
+          for service in ${{ steps.parse-services.outputs.services }}; do
+            echo -e "  ${service}:\n    image: ${service}:ci" >> docker-compose.ci.yaml
+          done
+          echo $ENCODED_CREDENTIALS | base64 -di > vault/aws_credentials
+
+      # TODO: This is a hack to place empty secret files in the vault directory
+      # (to satisfy docker-compose.yml requirements) without actually retrieving
+      # them from vault.
+      - name: Retrieve vault secrets
+        shell: bash
+        run: |
+          touch vault/website.secrets.env
+          touch vault/landingpageadmin.secrets.env
+
+      - name: Create service databases and roles
+        shell: bash
+        run: |
+          # Parse unique usernames, passwords, and DBs from POSTGRES_CONNECTION strings.
+          declare -A db_info
+          while read -r line; do
+            regex='postgres://(\w*):(\w*)@\$\{POSTGRES_SERVER\}/(\w*)'
+            if [[ ${line} =~ ${regex} ]]; then
+              # Map username to the associated password and DB.
+              db_info[${BASH_REMATCH[1]}]=${BASH_REMATCH[@]:2:2}
+            fi
+          done < <(grep POSTGRES_CONNECTION docker-compose.yml)
+          for username in "${!db_info[@]}"; do
+            read -r password db <<<$(echo ${db_info[${username}]})
+            # Start postgres in docker and create database and role.
+            docker-compose up -d postgres
+            docker-compose logs postgres
+            docker exec -i -e PGPASSWORD=${POSTGRES_PASSWORD} postgres psql -h 0.0.0.0 -U ${POSTGRES_USER} -c \
+              "CREATE DATABASE ${db}"
+            docker exec -i -e PGPASSWORD=${POSTGRES_PASSWORD} postgres psql -h 0.0.0.0 -U ${POSTGRES_USER} -c \
+              "CREATE ROLE ${username} LOGIN PASSWORD '${password}'"
+          done
+
+      - name: Launch services with dependencies via docker-compose
+        env:
+          REGISTRY: ${{ steps.login-ecr.outputs.registry }}
+        shell: bash
+        run: |
+          docker-compose pull --include-deps ${{ steps.parse-services.outputs.services }}
+          docker-compose -f docker-compose.yml -f docker-compose.ci.yaml \
+            up -d ${{ steps.parse-services.outputs.services }}
+
+      - name: View kafka logs, for debugging
+        shell: bash
+        run: |
+          docker-compose ps | grep kafka
+          docker-compose logs kafka
+
+      - name: View entity logs, for debugging
+        shell: bash
+        run: docker-compose logs entity
```

Comment:
Since we're implementing this workflow in `fact` now too, we should generalize this. I think this would work (unless you can't have variables in step names?):
```
      - name: View ${{ matrix.service }} logs, for debugging
        shell: bash
        run: docker-compose logs ${{ matrix.service }}
```

Reply:
Absolutely - that step is just for debugging and will be removed entirely before I put this up for final CR. There's kind of a lot of jank hanging around in this branch right now. (Pay no attention to the man behind the curtain! 🙈)